src/serializer/functions.js

```diff
@@ -57,26 +61,40 @@ export class Functions {
   writeEffects: WriteEffects;
   _noopFunction: void | ECMAScriptSourceFunctionValue;
 
+  _unwrapAbstract(value: AbstractValue): Value {
+    let elements = value.values.getElements();
+    if (elements) {
+      let possibleValues = [...elements].filter(
+        element => !(element instanceof EmptyValue || element instanceof UndefinedValue)
+      );
+      if (possibleValues.length === 1) {
+        return possibleValues[0];
+      }
+    }
+    return value;
+  }
+
   __optimizedFunctionEntryOfValue(value: Value): AdditionalFunctionEntry | void {
     let realm = this.realm;
-
+    // if we conditionally called __optimize, we may have an AbstractValue that is the union of Empty or Undefined and
+    // a function/component to optimize
     if (value instanceof AbstractValue) {
-      // if we conditionally called __optimize, we may have an AbstractValue that is the union of Empty or Undefined and
-      // a function/component to optimize
-      let elements = value.values.getElements();
-      if (elements) {
-        let possibleValues = [...elements].filter(
-          element => !(element instanceof EmptyValue || element instanceof UndefinedValue)
-        );
-        if (possibleValues.length === 1) {
-          value = possibleValues[0];
-        }
-      }
+      value = this._unwrapAbstract(value);
     }
-    if (value instanceof ECMAScriptSourceFunctionValue) {
-      // additional function logic
-      return { value };
-    } else if (value instanceof ObjectValue) {
+    invariant(value instanceof ObjectValue);
+    let funcValue = Get(realm, value, "funcValue");
+    if (funcValue !== realm.intrinsics.undefined) {
+      // unwrap from this side
```

Comment:
This double unwrapping was needed after[ this change](https://github.com/facebook/prepack/pull/2214). As a consequence here it is possible to get my ObjectValue wrapped in AbstractValue (union) or/and to get my ObjectValue having an AbstractValue (union) with function value. I haven't unwrapped anything in `__optimize` implementation because I'm not sure that it is safe there. Is it ok to do this in such a way? Or is it better to unwrap function value directly in `__optimize` implementation?